### DISCUSSION TITLE
Are there IDE completion stubs available?

### DISCUSSION THREAD

**Discussion by sewbacca** [chrisant996/clink#471] [Q&A] [answered]
I'd like to do some scripting and hope to get some autocompletion for scripts in VSCode. I am using [LuaLS](https://github.com/luals/lua-language-server). Are there any definition files available or maybe some formatted data, from which I can generate some definition files on my own?

**Comment by chrisant996** [comment]
Can you describe more about what you're looking for?

Are you asking for autocompletion in VSCode while editing Lua scripts in VSCode?

**Reply by sewbacca** [reply]
Exactly. As a workarround I scanned the environment and dumped the definitions into a normal Lua file which can be read by the language server, so I have at least some completion.
However what is missing are parameters and return values. The first could be extracted if these are native Lua functions, the latter can't be.
What I would need is something similar to these [serialized love2d definitions](https://github.com/love2d-community/love-api/blob/master/love_api.lua) or actual Lua files containing the definitions.
Since there is a documentation for the Lua reference, I could just parse the html files, which I think I'll do, if there is no other option, but it seems to me they are auto generated from something easier to parse.

**Reply by chrisant996** [reply, marked as answer]
> > Are you asking for autocompletion in VSCode while editing Lua scripts in VSCode?
>
> Exactly.

I can't tell which of these you mean:
- VSCode should assume that all Lua scripts will be executed by Clink, and always add completions for Clink's extensions to Lua.
- VSCode should somehow predict whether a Lua script is going to be executed by Clink, and if so then add completions for Clink's extensions to Lua.

Anyway, it sounds like the only part you're actually asking about is how to find what Lua functions Clink adds to standard Lua.
The HTML documentation for Clink is parsed from the Clink source code.
See [do_docs()](https://github.com/chrisant996/clink/blob/ab7b0416de8545160412a40667885ac7b64c4d32/docs/premake5.lua#L225) in the docs/premake5.lua script.

**Reply by sewbacca** [reply]
Thanks, that is what I was looking for! I just wanted autocompletion, but there are ways to make the language server autodetect `clink`. I almost finished a definition file generator, but am now wondering, whether or not I should make a pull request or rather keep the generator in a custom repository using this repo as a submodule. I guess it boils down to whether or not you are interested in providing IDE autocompletion files. What do you think?

**Reply by chrisant996** [reply]
It sounds like a separate repo would be appropriate.

**Reply by chrisant996** [reply]
@sewbacca Were you able to come up working VSCode autocompletions for Clink's Lua extensions?

If so, I'd be interested to see them.

Depending on what's involved, I might even be interested in auto-generating VSCode autocompletion files and including them in the Clink distribution itself.

**Reply by sewbacca** [reply]
Sadly no. I switched soon after almost exclusively to MSYS2, when it came to shells, which solved a lot of my problems, sorry =/

**Reply by sewbacca** [reply]
@chrisant996 actually scratch that. I just scanned my projects folder and did find a generated Lua file. See [sewbacca/clink](https://github.com/sewbacca/clink) for details. Shall I make a PR or do you just wanna grab what you need?

**Reply by chrisant996** [reply]
@sewbacca Thanks! I'm fine to grab stuff, I don't think a PR is needed. 

But can you point me at general info (or specific info!) for how to set up a Lua completion server (or whatever?) in VSCode?

**Reply by sewbacca** [reply]
Just install the Lua extension from VSCode and you'll be fine. Take a look at their [starting page](https://github.com/LuaLS/lua-language-server). There's a lot you can configure in `.luarc.json`. i.e. require paths are setup via `runtime.path` I think. VSCode can also load settings from `settings.json` when you prefix them with `Lua.`, but LuaLS is for all IDEs and they may not understand VSCodes `settings.json` (only important for committing I guess).
